_posts/2019-04-06-omakase-relay.markdown

```diff
@@ -0,0 +1,200 @@
+
+```ts
+import React from "react"
+import { createFragmentContainer, graphql } from "react-relay"
+import { MyProfile_me } from "__generated__/MyProfile_me.graphql"
+
+// [...]
+
+interface Props extends ViewProperties {
+  me: MyProfile_me
+}
+
+export class MyProfile extends React.Component<Props> {
+  render() {
+    return (
+      <View>
+        <Header>
+          <ProfilePhoto initials={props.me.initials} image={props.me.image} />
+          <Subheading>{props.me.name}</Subheading>
+        </Header>
+        <ButtonSection>
+          <ProfileButton
+            section="Selling"
+            description="Sell works from your collection"
+            onPress={startSubmission}
+          />
+          <ProfileButton
+            section="Account Details"
+            description="Email, password reset, profile"
+            onPress={goToUserSettings}
+          />
+        </ButtonSection>
+      </View>
+    )
+  }
+}
+
+export default createFragmentContainer(
+  MyProfile,
+  graphql`
+    fragment MyProfile_me on Me {
+      name
+      image
+      initials
+    }
+  `
+)
+```
+
+There are three moving parts:
+
+- The `MyProfile` component, which is a vanilla React component
+- The exported `createFragmentContainer` which wraps the `MyProfile` and ties it to a fragment on a `Me` type in
+  GraphQL.
+- The TypeScript interface `MyProfile_me` which ensures we have a correct interface to our props
+
+## Isolation
+
+The React component `MyProfile` will be passed in props that directly tie to the fragment that was requested. In
+Relay terms, this is called data-masking and it is one of the first hurdles for someone new to Relay to [grok][].
+In REST clients, and GraphQL API clients like Apollo Client, you make a request and that request is passed through
+the React tree. E.g.
+
+{% include epic_img.html url="/images/omakase-relay/tree.png" title="REST inspired props" style="width:100%;" %}
+
+This means most components know more about the request than it probably needs, as it may be needed to pass on to
+the component's children. This can lead to data-duplication, or even worse, not knowing if you can delete or
+refactor a component.
+
+Data masking solves this by hiding data that the component didn't request. I've still yet to find the right visual
+abstraction, but I feel this just about pays for itself.
+
+{% include epic_img.html url="/images/omakase-relay/isolation.png" title="Relay isolation tree" style="width:100%;" %}
+
+You let Relay be responsible for consolidating all your fragments into a query, request it, and (mostly) passing it
+through your component hierarchy. This means Relay powered component can be safely changed and drastically reduces
+the chance for unintended consequences elsewhere.
+
+This isolation gives Artsy engineers the safety to work on projects with tens of contributors over long time
+periods without too much technical debt. The components we create are nearly all focused only on the data-driven
```

Comment:
I wish there was a better / more elaborated way to express just how sturdy, and just how little technical debt (at least at the data layer) there typically is.